src/nodes.cpp

```diff
@@ -8,71 +8,117 @@
 #include <bbp/sonata/nodes.h>
 
-
 namespace bbp {
 namespace sonata {
 
-//--------------------------------------------------------------------------------------------------
-
-NodePopulation::NodePopulation(const std::string& h5FilePath,
-                               const std::string& csvFilePath,
-                               const std::string& name)
-    : Population(h5FilePath, csvFilePath, name, ELEMENT) {}
-
-
 namespace {
 
 template <typename T>
-Selection _getMatchingSelection(const std::vector<T>& values, const std::vector<T>& wanted) {
-    Selection::Values idx;
-    Selection::Value id = 0;
-
-    if (wanted.size() == 1) {
-        for (const auto& v : values) {
-            if (v == wanted[0]) {
-                idx.push_back(id);
-            }
-            ++id;
-        }
+Selection _matchAttributeValues(const NodePopulation& population,
+                                const std::string& name,
+                                const std::vector<T>& wanted) {
+    if (wanted.empty()) {
+        return Selection({});
+    } else if (wanted.size() == 1) {
+        return population.filterAttribute<T>(name,
+                                             [&wanted](const T& v) { return wanted[0] == v; });
     } else {
         std::vector<T> wanted_sorted(wanted);
         std::sort(wanted_sorted.begin(), wanted_sorted.end());
-        for (const auto& v : values) {
-            if (std::binary_search(wanted_sorted.cbegin(), wanted_sorted.cend(), v)) {
-                idx.push_back(id);
+
+        return population.filterAttribute<T>(name, [&wanted_sorted](const T& v) {
+            return std::binary_search(wanted_sorted.cbegin(), wanted_sorted.cend(), v);
+        });
+    }
+}
+
+bool is_unsigned_int(HighFive::DataType dtype) {
+    return dtype == HighFive::AtomicType<uint8_t>() || dtype == HighFive::AtomicType<uint16_t>() ||
+           dtype == HighFive::AtomicType<uint32_t>() || dtype == HighFive::AtomicType<uint64_t>();
+}
+
+bool is_signed_int(HighFive::DataType dtype) {
+    return dtype == HighFive::AtomicType<int8_t>() || dtype == HighFive::AtomicType<int16_t>() ||
+           dtype == HighFive::AtomicType<int32_t>() || dtype == HighFive::AtomicType<int64_t>();
+}
+bool is_floating(HighFive::DataType dtype) {
+    return dtype == HighFive::AtomicType<float>() || dtype == HighFive::AtomicType<double>();
+}
+
+template <typename UnaryPredicate>
+Selection _filterStringAttribute(const NodePopulation& population,
+                                 std::string name,
+                                 UnaryPredicate pred) {
+    if (population.enumerationNames().count(name) > 0) {
+        const auto& enum_values = population.enumerationValues(name);
+        std::vector<size_t> wanted_enum_value;
+        wanted_enum_value.reserve(enum_values.size());
+
+        for (size_t i = 0; i < enum_values.size(); ++i) {
+            if (pred(enum_values[i])) {
+                wanted_enum_value.push_back(i);
```

Comment:
`wanted_enum_value` could just be a LUT, `std::vector<bool>` or this bitvector-thingie.
```suggestion
                wanted_enum_value[i] = pred(enum_values[i]);
```
Then you'd have O(1) lookup speed at negligible memory-overhead, and ditch the search later.

Reply:
Hmm, yeah that could be interesting.